### ISSUE TITLE
Run snap with external mongo database

### ISSUE THREAD

**Issue by Yavor-Ivanov** [RocketChat/Rocket.Chat#14522]
I am trying to run the rocket chat snap with an already existing mongodb server. My reason for wanting this is that I have various other applications on my server, all of which already connect to a single mongodb instance. This setup simplifies backups and database administration, because I don't have to keep track of separate mongodb databases for every application.

I would like some way to configure rocketchat to use the mongo that is already running on my system instead of spinning up a separate database all for itself.

I have tried setting the mongo host and port for the snap to the one all other applications use, then stopping the `snap.rocketchat-server.rocketchat-mongo` service  and restarting the `snap.rocketchat-server.rocketchat-server` service. I had hoped that this will force the rocketchat server will connect to my non-snap mongo database, but I get the following error when I query the status:
```
$ sudo systemctl status snap.rocketchat-server.rocketchat-server

● snap.rocketchat-server.rocketchat-server.service - Service for snap application rocketchat-server.rocketchat-server
   Loaded: loaded (/etc/systemd/system/snap.rocketchat-server.rocketchat-server.service; enabled; vendor preset: disabled)
   Active: active (running) since Thu 2019-05-16 09:44:50 EEST; 1min 37s ago
 Main PID: 26704 (startRocketChat)
    Tasks: 2 (limit: 4915)
   Memory: 1.9M
   CGroup: /system.slice/snap.rocketchat-server.rocketchat-server.service
           ├─26704 /bin/bash /snap/rocketchat-server/1377/bin/startRocketChat
           └─29502 sleep 30

May 16 09:44:50 server-hostname-omitted systemd[1]: Stopped Service for snap application rocketchat-server.rocketchat-server.
May 16 09:44:50 server-hostname-omitted systemd[1]: Started Service for snap application rocketchat-server.rocketchat-server.
May 16 09:44:50 server-hostname-omitted rocketchat-server.rocketchat-server[26704]: Mongo is not available, can't start. Waiting 10 seconds and trying again
May 16 09:45:00 server-hostname-omitted rocketchat-server.rocketchat-server[26704]: Mongo is not available, can't start. Waiting 15 seconds and trying again
May 16 09:45:15 server-hostname-omitted rocketchat-server.rocketchat-server[26704]: Mongo is not available, can't start. Waiting 20 seconds and trying again
```

Is there any way to force snap.rocketchat-server to disable the internal mongodb service and use the one already running on the host?

**Comment by LuluGO** [comment]
Hello, there is a way to configure MONGO_URL and MONGO_OPLOG_URL connection strings to point to your non snap mongodb service, but you will have to leave snap.rocketchat-server.rocketchat-mongo service running, the start script for rocket.chat service checks availability for that mongodb service in order to start it, we are hoping to change that for futures snaps.
Depending on the complexity of your connection string, you can use one of these to methods to configure MONGO_URL and MONGO_OPLOG_URL

https://rocket.chat/docs/installation/manual-installation/ubuntu/snaps/#how-do-i-change-rocketchat-port-mongo_url-and-mongo_oplog_url-in-my-snap

https://rocket.chat/docs/installation/manual-installation/ubuntu/snaps/#how-do-i-change-other-environmental-variables-in-my-snap

**Comment by Yavor-Ivanov** [comment]
From what I can tell by reading the documentation, the `mongo-url` parameter affects where the `rocketchat-server` service connects to.

If the plan is to let the snap mongo instance run in addition to the non-snap mongodb instance I have, then the ports would have to be different, otherwise whichever mongo service runs first will claim the port, and the other will die off with an error.

I would be happy to let the snap mongodb instance run if I can configure it to run on a non-default port, but I didn't find anything about this in the rocketchat snap documentation. Is there a way to do this? I don't want to have to change the default port on my non-snap mongodb and reconfigure every application that uses it.

**Comment by LuluGO** [comment]
You are right about ports, it works where you want to use an external db server, you can't change the default port for the snap mongodb service, you can consider running snap rocketchat-server from another server that is not your db server, or maybe use another install method?

**Comment by Yavor-Ivanov** [comment]
I tried installing manually before going with the snap image, but I ran into some trouble with npm dependencies on Arch. The instance I'm running is a migration from a Debian manual installation that had no trouble with dependencies. I didn't want to invest any more time into trying to get the server working, so I ended up going with the snap image and a very hacky solution, which I'll detail below for anyone else interested in running such a setup:

Because rocket-chat server only ever checks that the snap mongo instance is running at start, I let the snap mongo run, then shut the service down and start my own mongo container. I put the below script on cron to check if rocketchat goes down (when snap decides to auto-update) and restart: 

```sh
restart_chat_server() {
        sudo systemctl stop mongodb
        sudo systemctl stop snap.rocketchat-server.rocketchat-mongo
        sudo systemctl stop snap.rocketchat-server.rocketchat-server
        sleep 30
        sudo systemctl restart snap.rocketchat-server.rocketchat-mongo
        sleep 30
        sudo systemctl restart snap.rocketchat-server.rocketchat-server
        sleep 30
        sudo systemctl stop snap.rocketchat-server.rocketchat-mongo
        sudo systemctl start mongodb
}

http_status="$(curl -s -o /dev/null -w "%{http_code}" chat.server-host-omitted.internal)"
if [ "$http_status" == 502 ]; then
        restart_chat_server
fi
```

It's not a very good solution, but at least I can run rocketchat on the same server, even if I do get a few minutes of downtime when an update happens.

**Comment by reetp** [comment]
Just run docker??
Snaps are fine for a quick & easy install, but not if you want to try something more exotic.
Docker is probably your best bet.
